lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,459 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+        /// Case WaitingForEvaluator:
+        ///    Adds Active Context to Active Context Manager
+        /// Case Fail:
+        ///    Closes the ActiveContext
+        /// Other cases - not expected
         /// </summary>
-        /// <param name="completedTask">The link to the completed task</param>
-        public void OnNext(ICompletedTask completedTask)
+        /// <param name="activeContext"></param>
+        public void OnNext(IActiveContext activeContext)
         {
-            lock (_completedTasks)
+            Logger.Log(Level.Verbose, "Received Active Context {0}, systemState {1}.", activeContext.Id, _systemState.CurrentState);
+            lock (_lock)
             {
-                Logger.Log(Level.Info,
-                    string.Format("Received completed task message from task Id: {0}", completedTask.Id));
-                _completedTasks.Add(completedTask);
-
-                if (AreIMRUTasksCompleted())
+                switch (_systemState.CurrentState)
                 {
-                    ShutDownAllEvaluators();
+                    case SystemState.WaitingForEvaluator:
+                        _contextManager.Add(activeContext);
+                        break;
+                    case SystemState.Fail:
+                        Logger.Log(Level.Info, "Received IActiveContext event, but system is in FAIL state. Closing the context.");
+                        activeContext.Dispose();
+                        break;
+                    default:
+                        UnexpectedState(activeContext.Id, "IActiveContext");
+                        break;
                 }
             }
         }
+        #endregion IActiveContext
 
+        #region submit tasks
         /// <summary>
-        /// Specifies what to do when evaluator fails.
-        /// If we get all completed tasks then ignore the failure
-        /// Else request a new evaluator. If failure happens in middle of IMRU 
-        /// job we expect neighboring evaluators to fail while doing 
-        /// communication and will use FailedTask and FailedContext logic to 
-        /// order shutdown.
+        /// Called from ActiveContextManager when all the expected active context are received.
+        /// It changes the system state then calls SubmitTasks().
         /// </summary>
         /// <param name="value"></param>
-        public void OnNext(IFailedEvaluator value)
+        public void OnNext(IAllContextsReceived value)
         {
-            if (AreIMRUTasksCompleted())
+            Logger.Log(Level.Info, "Received event from ActiveContextManager with NumberOfActiveContexts:" + value.GetNumberOfActiveContext());
+            lock (_lock)
             {
-                Logger.Log(Level.Info,
-                    string.Format("Evaluator with Id: {0} failed but IMRU task is completed. So ignoring.", value.Id));
-                return;
+                // Change the system state to SubmittingTasks
```

Comment:
This comment is confusing. Are we changing to `SubmittingTasks` or `AllContextsAreReady`?